src/cockpit/safely.py

```diff
@@ -0,0 +1,74 @@
+# This file is part of Cockpit.
+#
+# Copyright (C) 2023 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import contextlib
+
+from typing import Callable, Dict, Generator, List, Mapping, Type, TypeVar, Union
+
+T = TypeVar('T')
+D = TypeVar('D')
+
+
+@contextlib.contextmanager
+def catch(description: str, on_error: Callable) -> Generator[None, None, None]:
+    try:
+        yield
+    except (TypeError, KeyError) as err:
+        on_error(f'{description}: {err}')
+        raise
+
+
+def get(mapping: Mapping[str, object], key: str, expected_type: Type[T], *default: D) -> Union[T, D]:
```

Comment:
Admittedly I don't understand the syntactical difference between `Type[T]` and `D` here. But it seems dubious to me that the default can be of a different type than `expected_type`. We don't use that anywhere in packages.py.

I'd much prefer `default: Optional[T]` and return an `Optional[T]`, to support `None` values. That should give  better type checking and avoid bugs.

Same for get_{list,dict} below.

Reply:
It's true that the only usecase I had in mind here was for `None`, and indeed that could be better covered by `Optional[T]`, but: in the case that the default value is not given, or the default value is something other than `None`, then this would cause the return type of this function to be something like `Optional[str]` when it could be `str`.

I struggle with the right thing to do here...

Reply:
Hmm, but if no `default` is given, the return type would be `Union[T, D]` with `D` being entirely undefined. Unless typing/mypy is clever enough to realize that there is no `D` and it would collapse to a `Union[T]` then? But this is still murky for the "given default" case -- being able to call it with `(map, "key", int, "foo")` just feels broken.

If you care for a strict return type, perhaps have a `.get()` which would `KeyError`, and a `get_default()` with a mandatory `default` argument?